bpfd-operator/controllers/bpfd-agent/common.go

```diff
@@ -234,109 +235,120 @@ func (r *ReconcilerCommon) createBpfProgram(ctx context.Context,
+			if updatedFinalizers {
 				return false, nil
 			}
-			return r.updateStatus(ctx, &bpfProg, progCond)
-		}
 
-		// Make sure if we're not selected exit and write correct condition
-		if !isNodeSelected {
-			r.Logger.V(1).Info("Program does not select this node")
-			// Write NodeNodeSelected status
-			return r.updateStatus(ctx, &bpfProg, progCond)
+			updatedStatus := r.updateStatus(ctx, &bpfProg, progCond)
+			if updatedStatus {
+				return false, nil
+			}
 		}
+	// If the *Program isn't being deleted ALWAYS create the bpfPrograms
+	// even if the node isn't selected
+	case false:
+		for _, expectedProg := range expectedPrograms.Items {
+			prog, exists := existingPrograms[expectedProg.Name]
+			if !exists {
+				opts := client.CreateOptions{}
+				r.Logger.Info("creating bpfProgram", "Name", expectedProg.Name, "Owner", program.GetName())
+				if err := r.Create(ctx, &expectedProg, &opts); err != nil {
+					return true, fmt.Errorf("failed to create bpfProgram object: %v", err)
+				}
+				existingPrograms[expectedProg.Name] = prog
+				return false, nil
+			}
 
-		// If bpfProgram Maps isn't up to date just update it and return
-		if !reflect.DeepEqual(bpfProg.Spec.Maps, r.expectedMaps) && len(r.expectedMaps) != 0 {
-			r.Logger.V(1).Info("Updating bpfProgram Object", "Maps", r.expectedMaps, "bpfProgram", bpfProg.Name)
-			bpfProg.Spec.Maps = r.expectedMaps
-			if err := r.Update(ctx, &bpfProg, &client.UpdateOptions{}); err != nil {
-				r.Logger.Error(err, "failed to update bpfProgram's Programs")
-				return true, nil
+			// bpfProgram Object exists go ahead and reconcile it.
+			cond, err := rec.reconcileBpfdProgram(ctx, programMap, bytecode, &prog, isNodeSelected, isBeingDeleted)
+			if err != nil {
+				r.Logger.Error(err, "Failed to reconcile bpfd")
 			}
-			return false, nil
-		}
 
-		alreadySet, err := r.updateStatus(ctx, &bpfProg, progCond)
-		if err != nil {
-			r.Logger.Error(err, "failed to update bpfProgram's Status")
-			return true, nil
-		}
+			// Make sure if we're not selected exit and write correct condition
+			if cond == bpfdiov1alpha1.BpfProgCondNotSelected {
+				r.Logger.V(1).Info("Program does not select this node")
+				// Write NodeNodeSelected status
+				updatedStatus := r.updateStatus(ctx, &prog, cond)
+				if updatedStatus {
+					return false, nil
+				}
+			}
 
-		// Only stop reconcile loop if we updated the status for a bpfprogram
-		if !alreadySet {
-			return false, nil
+			// If bpfProgram Maps isn't up to date just update it and return
+			if !reflect.DeepEqual(prog.Spec.Maps, r.expectedMaps) && len(r.expectedMaps) != 0 {
+				r.Logger.V(1).Info("Updating bpfProgram Object", "Maps", r.expectedMaps, "bpfProgram", prog.Name)
+				prog.Spec.Maps = r.expectedMaps
+				if err := r.Update(ctx, &prog, &client.UpdateOptions{}); err != nil {
+					r.Logger.Error(err, "failed to update bpfProgram's Programs")
+					return true, nil
```

Comment:
Confused on the `retry` flag. Here `return true, nil`, but most everywhere else, it is either `return false, nil` or `return true, fmt.Errorf()`.

In `Reconcile()` in xdp-program.go
```
func Reconcile() {
	:
	for _, xdpProgram := range xdpPrograms.Items {
		:
		retry, err := reconcileProgram(ctx, r, r.currentXdpProgram, &r.currentXdpProgram.Spec.BpfProgramCommon, r.ourNode, programMap)
		if err != nil {
			r.Logger.Error(err, "Reconciling XdpProgram Failed", "XdpProgramName", r.currentXdpProgram.Name, "Retrying", retry)
			return ctrl.Result{Requeue: retry, RequeueAfter: retryDurationAgent}, nil
		}
	}
	return ctrl.Result{Requeue: false}, nil
}
```
So retry flag is only honored in there is an error returned.

Reply:
Hrm good point I think this is broken... 

Basically what I wanted to do was ensure that we NEVER return and error for `Reconcile` because it will trigger a TON of instant reconciles and will not honor the retry period...basically just infinite looping most of the time :/ .

What this should express is the opposite... If we return an Error, print it and retry with the default retry duration (5 sec)

Let me try and clean this up in a new commit.

Reply:
PTAL at 7b71886fb1a03aa1cace7fdc1b0a44c9499a5933